File: src/main/java/ru/sejapoe/digitalhotelserver/core/security/BitArray256.java
package ru.sejapoe.digitalhotelserver.core.security;

import org.springframework.data.mongodb.core.mapping.Document;

import java.math.BigInteger;
import java.util.Base64;

@Document
public final class BitArray256 {
    private byte[] arr;

    public BitArray256() {
        arr = new byte[32];
    }

    private BitArray256(byte[] bytes) {
        arr = bytes;
    }

    public static BitArray256 fromBigInteger(BigInteger source) {
        return fromByteArray(source.toByteArray());
    }

    public static BitArray256 fromByteArray(byte[] source) {
        byte[] bytes = new byte[32];
        if (source.length != 0) {
            System.arraycopy(source, 0, bytes, Integer.max(32 - source.length, 0), Integer.min(source.length, 32));
        }
        return new BitArray256(bytes);
    }

    public static BitArray256 fromBase64(String source) {
        return fromByteArray(Base64.getDecoder().decode(source));
    }

    public byte[] asByteArray() {
        return arr;
    }

    public BigInteger asBigInteger() {
        StringBuilder result = new StringBuilder();
        for (byte b : arr) {
            result.append(Integer.toHexString(b));
        }
        return new BigInteger(result.toString(), 16);
    }

    public String asBase64() {
        return Base64.getEncoder().encodeToString(arr);
    }
}

File: src/jvmMain/kotlin/ru/sejapoe/application/utils/BitArray256.kt
package ru.sejapoe.application.utils

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.math.BigInteger
import java.util.*

@Serializable
data class BitArray256(
    val bytes: ByteArray
) {
    fun asByteArray(): ByteArray {
        return bytes
    }

    fun asBigInteger(): BigInteger {
        val result = StringBuilder()
        for (b in bytes) {
            result.append(Integer.toHexString(b.toInt()))
        }
        return BigInteger(result.toString(), 16)
    }

    fun asBase64(): String {
        return Base64.getEncoder().encodeToString(bytes)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as BitArray256

        return bytes.contentEquals(other.bytes)
    }

    override fun hashCode(): Int {
        return bytes.contentHashCode()
    }

    companion object {
        fun fromBigInteger(source: BigInteger): BitArray256 {
            return fromByteArray(source.toByteArray())
        }

        fun fromByteArray(source: ByteArray): BitArray256 {
            val bytes = ByteArray(32)
            if (source.isNotEmpty()) {
                System.arraycopy(source, 0, bytes, Integer.max(32 - source.size, 0), Integer.min(source.size, 32))
            }
            return BitArray256(bytes)
        }

        fun fromBase64(source: String?): BitArray256 {
            return fromByteArray(Base64.getDecoder().decode(source))
        }
    }
}
